### PR TITLE
Cache and interpolate resolutions when drawing sample waveforms

### PR DESCRIPTION
**Pull request by sakertooth**
This PR is meant to improve performance when drawing sample waveforms by:
1. Caching multiple waveform resolutions that are decimated in logarithms (currently by powers of 2)
2. Interpolating between the cached waveform resolutions to get the resolution necessary for the current zoom level (currently done using linear interpolation)

That's about it. I do want to note that when stretching sample clips vertically and using continuous scrolling, there is still a major drop in performance due to the drawing that's being done on the CPU. In the future, we should migrate such drawing to be done by the GPU using something like OpenGL.

Below is an attachment for the results of `perf` when sampling cycles in the scenario stated above: as shown, most of the overhead in this case lies within the Qt library itself, more specifically its software rasterization that is done on the CPU.
image

I have also omitted RMS calculations from the waveform visualization since most DAWs (and now even Audacity when using default settings) don't consider this, and I wanted to prioritize more performance.

### CODE REVIEW

**Comment by khoidauminh** [comment]
Pretty decent performance! Though I'm seeing some bugs

The sample zoom jitter is still there:

https://github.com/user-attachments/assets/0fc81c40-4b5d-460e-8d6e-c5b629395b8e

Waveform not being drawn in SlicerT (or you haven't drawn it yet):

ảnh

Incorrect rendering in AFP:

https://github.com/user-attachments/assets/f8a4765f-dd0e-415a-b0fb-7a976df7b30d

Also iirc, this one also uses more ram, but it's not a big problem.

Have you tested how this compares to my implementation?

**Comment by khoidauminh** [comment]
If you don't mind, I'll push some fixes

**Comment by sakertooth** [comment]
> If you don't mind, I'll push some fixes

Yeah, that's fine with me.

File: src/gui/SampleWaveform.cpp [hunk 1]
```diff
@@ -23,73 +23,93 @@
...
+	std::optional<size_t> from, std::optional<size_t> to)
+{
+	if (!m_buffer || m_size == 0) { return; }
 
-	size_t numPixels = std::min(parameters.size, static_cast<size_t>(width));
-	auto min = std::vector<float>(numPixels, 1);
-	auto max = std::vector<float>(numPixels, -1);
-	auto squared = std::vector<float>(numPixels, 0);
+	const auto sampleBegin = from.has_value() ? from.value() : 0;
+	const auto sampleEnd = to.has_value() ? to.value() : m_size;
```
Comment (messmerd) [inline src/gui/SampleWaveform.cpp:76]:
```suggestion
	const auto sampleBegin = from.value_or(0);
	const auto sampleEnd = to.value_or(m_size);
```

Comment (sakertooth) [reply src/gui/SampleWaveform.cpp:76]: Oh yeah, I told myself I was going to make this change but I completely forgot, thanks.

**Comment by sakertooth** [comment]
@khoidauminh, I tested your PR. I don't remember the implementation being that good, wow. I think I might close this PR actually and continue contributing to what you have over there. And the ideas we did here are the same, at least in terms of caching, but I'm shocked at how the continuous scrolling is that fast with your PR.

**Comment by khoidauminh** [comment]
@sakertooth Damn thank you

Though I think my PR is only faster because I clipped rendering to the visible region

This PR did make me realize how messy my code still is lol. I'll see what I can clean up

**Pull request closed without merging**